### ISSUE TITLE
[RSS 2.0] Unexpected Entry Updated

### ISSUE THREAD

**Issue by jangernert** [feed-rs/feed-rs#235]
This isn't really a bug report, but more of a question about what should be the expected behavior:

The feed in question: https://0pointer.net/blog/index.rss20 (as of creation of this issue: [rss2-feed.zip](https://github.com/user-attachments/files/16256517/rss2-feed.zip))

The feed has the property

```
<lastBuildDate>Wed, 12 Jun 2024 00:00:00 +0200</lastBuildDate>
```

indicating the last time the *feed* was changed. And this makes sense because the latest entry has

```
<pubDate>Wed, 12 Jun 2024 00:00:00 +0200</pubDate>
```

No entry in this feed has a property of `lastBuildDate`. So I would expect `feed.updated` to be `Wed, 12 Jun 2024 00:00:00 +0200` but `entry.updated` to be `None` on all entries.

For some reason all entries get assigned the updated timestamp of the feed: https://github.com/feed-rs/feed-rs/blob/master/feed-rs/src/parser/rss2/mod.rs#L76

Which means every time the feed updates Newsflash will think all entries were updated as well.
What was the idea behind this?

**Comment by markpritchard** [comment]
Hey @jangernert , 

The goal for feed-rs is to behave predictably across all of the different feed formats (Atom, RSS etc). 

Unfortunately, there isn't a concept of "updated" date for an item (entry) in RSS 2.0, and `lastBuildDate` isn't defined for entries (items) in the RSS 2.0 spec either.

So feed-rs tries to fill in the parsed model as best it can.

The corresponding code (from all the way back in 2019!)...
```rust
    // RSS 2.0 defines <lastBuildDate> on an item as optional so for completeness we set them to the updated date of the feed
    for entry in feed.entries.iter_mut() {
        entry.updated = feed.updated;
    }
```

The idea being that a feed (RSS channel) has been updated, and we assume the entries (RSS items) within that feed have been updated too from the model perspective.

Note that `entry.published` would be set correctly in this case.

Does that help explain the behaviour?

**Comment by jangernert** [comment]
Not really. I'd personally prefer feed-rs to not hallucinate information (with the exception of stable entry ids). `feed.updated` is still correctly filled and could be used as a fallback when `entry.updated` is not present. With the current behavior there is no way of knowing if `entry.updated` is "real" or just taken from the feed (apart from comparing values).

But that's just my 2ct. Feel free to close and ignore this and I'll try to work around it.

**Comment by markpritchard** [comment]
feed-rs does quite a few transformations (e.g. normalising mime types and content encoding) but I take your point about not creating data. My goal was to populate the `entry.updated` field with useful data, and if doing that with the `feed.updated` field isn't useful, or even worse, makes it confusing/broken for _users_ then perhaps it should change.

Atom requires the updated date, and JSONFeed supports it too. RSS2 is the dominant schema feed-rs sees, but it would be nice for users of the crate if they could rely on `entry.updated` being set to a useful value so all schemas are processed in a similar way for users of the crate.

What if `feed-rs` sets `entry.updated` to the source `item.pubDate` ? i.e. if present, `item.pubDate` would be parsed into `entry.published` _and_ `entry.updated`.

That way `feed-rs` isn't making up data, just continuing to transform it into a common representation.

**Comment by jangernert** [comment]
> feed-rs does quite a few transformations (e.g. normalising mime types and content encoding) but I take your point about not creating data. My goal was to populate the entry.updated field with useful data, and if doing that with the feed.updated field isn't useful, or even worse, makes it confusing/broken for users then perhaps it should change.

I'm just one user of your crate with one opinion. Maybe there is a way to get other opinions on this?

> What if feed-rs sets entry.updated to the source item.pubDate ? i.e. if present, item.pubDate would be parsed into entry.published and entry.updated.

That would work better for me. At least this wouldn't suggest that every entry was updated when only a single new was added to the feed.

But what I don't quite understand is: why does feed-rs need to try so hard to fill `entry.updated`? The docs very clearly state

> Time at which this item was last modified. If not provided in the source, or invalid, it is `None`.

So if there is no information I'd assume it just be `None`.

**Comment by markpritchard** [comment]
> But what I don't quite understand is: why does feed-rs need to try so hard to fill entry.updated? The docs very clearly state
Because I built `feed-rs` for my own use case. It uses the `entry.updated` field for re-indexing content in a search engine. It isn't as sensitive to this problem since its just a little extra work that wasn't required if the item is incorrectly flagged as updated.

I've never stated that `feed-rs` is perfect, nor that I'm an expert at the various schemas/standards so you've found a clear gap / bug in documentation and behaviour.

As I've mentioned previously, I'm happy to consider changing it. Populating it with the published date solves my problem too (understanding when an item has changed in a consistent way).

> I'm just one user of your crate with one opinion. Maybe there is a way to get other opinions on this?
I think one of the commentors on the serialisation issue nailed it - this crate is too "niche" to source much feedback on these types of changes. So I'm happy to kick around some ideas, and then change it with a minor version bump + clear doco. Perhaps that creates more bug reports, and then I'll iterate on things (e.g. parser flags etc).

**Comment by jangernert** [comment]
> I've never stated that feed-rs is perfect, nor that I'm an expert at the various schemas/standards

It's pretty close though :)

> It isn't as sensitive to this problem since its just a little extra work

It's not exactly a deal breaker for my use case as well, otherwise I would have gotten swarmed with issues. It's just something I've noticed debugging another issue. The same data gets written to the database over and over. And the feed I used as an example with its ~400 entries is of course an outlier. But it was a noticeable amount of data that was being written. It caught me off guard and made me dig for the cause.

**Comment by jangernert** [comment]
Thanks for the quick update. I finally had a chance to verify the changes: a lot less articles written to the database every refresh. Feels good!

**Comment by markpritchard** [comment]
Excellent news, thanks for letting me know @jangernert !
